{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ffd6f8f0-777f-4e53-905b-94ad0c4c4171",
   "metadata": {},
   "source": [
    "# Build background set of variants from transcripts genomewide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d9379bc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import random\n",
    "from datetime import date\n",
    "import matplotlib.pyplot as plt\n",
    "import itertools\n",
    "import csv\n",
    "from collections import Counter\n",
    "import pybedtools as pbt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d316697f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import splanl.post_processing as pp\n",
    "import splanl.custom_splai_scores as css\n",
    "import splanl.create_vcf as vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "98c26d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "annots_dir = '/nfs/kitzman2/smithcat/refs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8a4dfc12",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_annot = pd.read_table( annots_dir + 'gencode/gencode.v38lift37.MANESelect.gtf.gz',\n",
    "                               compression = 'gzip',\n",
    "                                sep = '\\t',\n",
    "                               comment = '#',\n",
    "                               index_col = False,\n",
    "                               names = [ 'chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute' ] )\n",
    "                                #dtype = { 'CHROM': object } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1ea25be2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>gene</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7_1\"; gene_type \"prot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>transcript</td>\n",
       "      <td>65419</td>\n",
       "      <td>71585</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>65419</td>\n",
       "      <td>65433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>65520</td>\n",
       "      <td>65573</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186092.7_1\"; transcript_id \"...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476742</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>59272371</td>\n",
       "      <td>59272463</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476743</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>exon</td>\n",
       "      <td>59274553</td>\n",
       "      <td>59276439</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476744</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>59274553</td>\n",
       "      <td>59274618</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476745</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>UTR</td>\n",
       "      <td>59274619</td>\n",
       "      <td>59276439</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476746</th>\n",
       "      <td>chrY</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>59274619</td>\n",
       "      <td>59274621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>476747 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom  source     feature     start       end score strand frame  \\\n",
       "0       chr1  HAVANA        gene     65419     71585     .      +     .   \n",
       "1       chr1  HAVANA  transcript     65419     71585     .      +     .   \n",
       "2       chr1  HAVANA        exon     65419     65433     .      +     .   \n",
       "3       chr1  HAVANA         UTR     65419     65433     .      +     .   \n",
       "4       chr1  HAVANA        exon     65520     65573     .      +     .   \n",
       "...      ...     ...         ...       ...       ...   ...    ...   ...   \n",
       "476742  chrY  HAVANA         CDS  59272371  59272463     .      +     0   \n",
       "476743  chrY  HAVANA        exon  59274553  59276439     .      +     .   \n",
       "476744  chrY  HAVANA         CDS  59274553  59274618     .      +     0   \n",
       "476745  chrY  HAVANA         UTR  59274619  59276439     .      +     .   \n",
       "476746  chrY  HAVANA  stop_codon  59274619  59274621     .      +     0   \n",
       "\n",
       "                                                attribute  \n",
       "0       gene_id \"ENSG00000186092.7_1\"; gene_type \"prot...  \n",
       "1       gene_id \"ENSG00000186092.7_1\"; transcript_id \"...  \n",
       "2       gene_id \"ENSG00000186092.7_1\"; transcript_id \"...  \n",
       "3       gene_id \"ENSG00000186092.7_1\"; transcript_id \"...  \n",
       "4       gene_id \"ENSG00000186092.7_1\"; transcript_id \"...  \n",
       "...                                                   ...  \n",
       "476742  gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...  \n",
       "476743  gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...  \n",
       "476744  gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...  \n",
       "476745  gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...  \n",
       "476746  gene_id \"ENSG00000124333.16_1_PAR_Y\"; transcri...  \n",
       "\n",
       "[476747 rows x 9 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_annot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4dd18f27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "476747"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( gencode_annot )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "7092b455",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_annot[ 'gene_type' ] = [ a.split( ' ' )[ 1 ].replace( '\"', '' ) \n",
    "                                 for att in gencode_annot.attribute \n",
    "                                 for a in att.split( '; ' )\n",
    "                                 if a.split( ' ' )[ 0 ] == 'gene_type' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6a9a8670",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_annot[ 'gene_id' ] = [ a.split( ' ' )[ 1 ].replace( '\"', '' ) \n",
    "                                 for att in gencode_annot.attribute \n",
    "                                 for a in att.split( '; ' )\n",
    "                                 if a.split( ' ' )[ 0 ] == 'gene_id' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "62e5177b",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_annot.loc[ gencode_annot.feature != 'gene', 'transcript_id' ] = [ a.split( ' ' )[ 1 ].replace( '\"', '' )\n",
    "                                                                          for att in gencode_annot.attribute\n",
    "                                                                          for a in att.split( '; ' )\n",
    "                                                                          if a.split( ' ' )[ 0 ] == 'transcript_id'  ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a3da521c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17631"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( gencode_annot.gene_id.unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b6b8e746",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17632"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#same length - the additional transcript is NaN from the gene rows\n",
    "len( gencode_annot.transcript_id.unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c5d01df8",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_annot.loc[ gencode_annot.feature == 'CDS', 'exon_id' ] = [ a.split( ' ' )[ 1 ].replace( '\"', '' )\n",
    "                                                                          for att in gencode_annot.loc[ gencode_annot.feature == 'CDS' ].attribute\n",
    "                                                                          for a in att.split( '; ' )\n",
    "                                                                          if a.split( ' ' )[ 0 ] == 'exon_id'  ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ccf5d93",
   "metadata": {},
   "outputs": [],
   "source": [
    "ex_counts = gencode_annot[ [ 'transcript_id', 'exon_id' ] ].groupby( 'transcript_id' ).nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "02aecc21",
   "metadata": {},
   "outputs": [],
   "source": [
    "#need at least 2 if I want to remove the first and last exons\n",
    "mult_exons = ex_counts.loc[ ( ex_counts.exon_id > 2 ) ].index.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bda3c2cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_mult = gencode_annot.set_index( 'transcript_id' ).loc[ mult_exons ].reset_index().copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "fb3910b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14747"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( gencode_mult.gene_id.unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "504dd2cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_prot = gencode_mult.loc[ gencode_mult.gene_type == 'protein_coding' ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "2d6b5518",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14747"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( gencode_prot.gene_id.unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "517d614a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_cds = gencode_prot.loc[ gencode_prot.feature == 'CDS' ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "fbba938e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14747"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( gencode_cds.gene_id.unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b998e6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only middle coding exons\n",
    "gencode_middle = gencode_cds.groupby( 'gene_id' ).apply( lambda group: group.iloc[ 1: -1 ] ).drop( columns = 'gene_id' ).reset_index().drop( columns = 'level_1' ).copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "29e273a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14747"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( gencode_middle.gene_id.unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "55e2e50b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_bed( gencode_gtf,\n",
    "                intron_bds = 100 ):\n",
    "    \n",
    "    gtf = gencode_gtf.copy()\n",
    "    \n",
    "    outbed = { 'chrom': [],\n",
    "               'start': [],\n",
    "               'end': [],\n",
    "               'name': [],\n",
    "               'score': [],\n",
    "               'strand': [] }\n",
    "    \n",
    "    for trans, trans_df in gtf.groupby( 'transcript_id' ):\n",
    "        \n",
    "        outbed[ 'chrom' ].append( trans_df.chrom.unique()[ 0 ] )\n",
    "        outbed[ 'start' ].append( int( trans_df.start.min() - intron_bds - 1 ) )\n",
    "        outbed[ 'end' ].append( int( trans_df.end.max() + intron_bds ) )\n",
    "        outbed[ 'name' ].append( trans )\n",
    "        outbed[ 'score' ].append( '.' )\n",
    "        outbed[ 'strand' ].append( trans_df.strand.unique()[ 0 ] )\n",
    "        \n",
    "    outdf = pd.DataFrame( outbed ).sort_values( by = [ 'chrom', 'start' ] )\n",
    "    \n",
    "    return outdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "55545a57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 5.94 s, sys: 0 ns, total: 5.94 s\n",
      "Wall time: 5.94 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "cds_bed_df = create_bed( gencode_middle )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "e54df1a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_bed = pbt.BedTool.from_dataframe( cds_bed_df )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "519f84df",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_int = cds_bed.intersect( cds_bed,\n",
    "                             wa = True,\n",
    "                             wb = True ).to_dataframe().rename( columns = { 'thickStart': 'chrom2',\n",
    "                                                                            'thickEnd': 'start2',\n",
    "                                                                            'itemRgb': 'end2',\n",
    "                                                                            'blockCount': 'name2',\n",
    "                                                                            'blockSizes': 'score2',\n",
    "                                                                            'blockStarts': 'strand2' } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cca8f0d8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>chrom2</th>\n",
       "      <th>start2</th>\n",
       "      <th>end2</th>\n",
       "      <th>name2</th>\n",
       "      <th>score2</th>\n",
       "      <th>strand2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>861201</td>\n",
       "      <td>879288</td>\n",
       "      <td>ENST00000616016.5_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>861201</td>\n",
       "      <td>879288</td>\n",
       "      <td>ENST00000616016.5_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>880336</td>\n",
       "      <td>894561</td>\n",
       "      <td>ENST00000327044.7_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>880336</td>\n",
       "      <td>894561</td>\n",
       "      <td>ENST00000327044.7_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>896572</td>\n",
       "      <td>900010</td>\n",
       "      <td>ENST00000338591.8_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>896572</td>\n",
       "      <td>900010</td>\n",
       "      <td>ENST00000338591.8_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>901983</td>\n",
       "      <td>909844</td>\n",
       "      <td>ENST00000379410.8_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>901983</td>\n",
       "      <td>909844</td>\n",
       "      <td>ENST00000379410.8_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>934805</td>\n",
       "      <td>935267</td>\n",
       "      <td>ENST00000304952.11_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1</td>\n",
       "      <td>934805</td>\n",
       "      <td>935267</td>\n",
       "      <td>ENST00000304952.11_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15438</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26774268</td>\n",
       "      <td>26777546</td>\n",
       "      <td>ENST00000382392.5_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26774268</td>\n",
       "      <td>26777546</td>\n",
       "      <td>ENST00000382392.5_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15439</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26919616</td>\n",
       "      <td>26952828</td>\n",
       "      <td>ENST00000382365.7_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26919616</td>\n",
       "      <td>26952828</td>\n",
       "      <td>ENST00000382365.7_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15440</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26986777</td>\n",
       "      <td>27042786</td>\n",
       "      <td>ENST00000682740.1_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>26986777</td>\n",
       "      <td>27042786</td>\n",
       "      <td>ENST00000682740.1_1</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15441</th>\n",
       "      <td>chrY</td>\n",
       "      <td>27184855</td>\n",
       "      <td>27188133</td>\n",
       "      <td>ENST00000382287.5_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>27184855</td>\n",
       "      <td>27188133</td>\n",
       "      <td>ENST00000382287.5_1</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15442</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59228191</td>\n",
       "      <td>59272563</td>\n",
       "      <td>ENST00000286448.12_1_PAR_Y</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>chrY</td>\n",
       "      <td>59228191</td>\n",
       "      <td>59272563</td>\n",
       "      <td>ENST00000286448.12_1_PAR_Y</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15443 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end                        name score strand  \\\n",
       "0      chr1    861201    879288         ENST00000616016.5_1     .      +   \n",
       "1      chr1    880336    894561         ENST00000327044.7_1     .      -   \n",
       "2      chr1    896572    900010         ENST00000338591.8_1     .      +   \n",
       "3      chr1    901983    909844         ENST00000379410.8_1     .      +   \n",
       "4      chr1    934805    935267        ENST00000304952.11_1     .      -   \n",
       "...     ...       ...       ...                         ...   ...    ...   \n",
       "15438  chrY  26774268  26777546         ENST00000382392.5_1     .      +   \n",
       "15439  chrY  26919616  26952828         ENST00000382365.7_1     .      -   \n",
       "15440  chrY  26986777  27042786         ENST00000682740.1_1     .      +   \n",
       "15441  chrY  27184855  27188133         ENST00000382287.5_1     .      -   \n",
       "15442  chrY  59228191  59272563  ENST00000286448.12_1_PAR_Y     .      +   \n",
       "\n",
       "      chrom2    start2      end2                       name2 score2 strand2  \n",
       "0       chr1    861201    879288         ENST00000616016.5_1      .       +  \n",
       "1       chr1    880336    894561         ENST00000327044.7_1      .       -  \n",
       "2       chr1    896572    900010         ENST00000338591.8_1      .       +  \n",
       "3       chr1    901983    909844         ENST00000379410.8_1      .       +  \n",
       "4       chr1    934805    935267        ENST00000304952.11_1      .       -  \n",
       "...      ...       ...       ...                         ...    ...     ...  \n",
       "15438   chrY  26774268  26777546         ENST00000382392.5_1      .       +  \n",
       "15439   chrY  26919616  26952828         ENST00000382365.7_1      .       -  \n",
       "15440   chrY  26986777  27042786         ENST00000682740.1_1      .       +  \n",
       "15441   chrY  27184855  27188133         ENST00000382287.5_1      .       -  \n",
       "15442   chrY  59228191  59272563  ENST00000286448.12_1_PAR_Y      .       +  \n",
       "\n",
       "[15443 rows x 12 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cds_int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "262d9bd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_int = cds_int.loc[ ( cds_int.name != cds_int.name2 ) ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "773b25ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "696"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( cds_int )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "5d947918",
   "metadata": {},
   "outputs": [],
   "source": [
    "int_trans = pd.concat( [ cds_int.name, cds_int.name2 ],\n",
    "                       axis = 0 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5f4b8aa5",
   "metadata": {},
   "outputs": [],
   "source": [
    "int_trans = int_trans.unique().tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "450548fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "nover_trans = gencode_middle.set_index( 'transcript_id' ).index.difference( int_trans )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "8df77a6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "gencode_nover = gencode_middle.set_index( 'transcript_id' ).loc[ nover_trans ].reset_index().copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d399a277",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14618"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( gencode_nover.gene_id.unique() )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7660229a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>chrom</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_type</th>\n",
       "      <th>exon_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>ENSG00000004059.11_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127229137</td>\n",
       "      <td>127229217</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000004059.11_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00003494180.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>ENSG00000004059.11_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127229539</td>\n",
       "      <td>127229648</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000004059.11_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00003504066.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>ENSG00000004059.11_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127230120</td>\n",
       "      <td>127230191</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000004059.11_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00003678978.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>ENSG00000004059.11_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>127231017</td>\n",
       "      <td>127231142</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000004059.11_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00003676786.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENST00000000442.11_1</td>\n",
       "      <td>ENSG00000173153.17_1</td>\n",
       "      <td>chr11</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>64081423</td>\n",
       "      <td>64081539</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000173153.17_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00003589560.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141812</th>\n",
       "      <td>ENST00000684770.1_1</td>\n",
       "      <td>ENSG00000170379.21_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>143416776</td>\n",
       "      <td>143417767</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>gene_id \"ENSG00000170379.21_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00003734225.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141813</th>\n",
       "      <td>ENST00000684770.1_1</td>\n",
       "      <td>ENSG00000170379.21_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>143418374</td>\n",
       "      <td>143418471</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000170379.21_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00002528833.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141814</th>\n",
       "      <td>ENST00000684770.1_1</td>\n",
       "      <td>ENSG00000170379.21_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>143419751</td>\n",
       "      <td>143419964</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000170379.21_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00002533605.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141815</th>\n",
       "      <td>ENST00000684770.1_1</td>\n",
       "      <td>ENSG00000170379.21_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>143420304</td>\n",
       "      <td>143420543</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000170379.21_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00002443125.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>141816</th>\n",
       "      <td>ENST00000684770.1_1</td>\n",
       "      <td>ENSG00000170379.21_1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>HAVANA</td>\n",
       "      <td>CDS</td>\n",
       "      <td>143421453</td>\n",
       "      <td>143421790</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000170379.21_1\"; transcript_id ...</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ENSE00002460631.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>141817 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               transcript_id               gene_id  chrom  source feature  \\\n",
       "0       ENST00000000233.10_1  ENSG00000004059.11_1   chr7  HAVANA     CDS   \n",
       "1       ENST00000000233.10_1  ENSG00000004059.11_1   chr7  HAVANA     CDS   \n",
       "2       ENST00000000233.10_1  ENSG00000004059.11_1   chr7  HAVANA     CDS   \n",
       "3       ENST00000000233.10_1  ENSG00000004059.11_1   chr7  HAVANA     CDS   \n",
       "4       ENST00000000442.11_1  ENSG00000173153.17_1  chr11  HAVANA     CDS   \n",
       "...                      ...                   ...    ...     ...     ...   \n",
       "141812   ENST00000684770.1_1  ENSG00000170379.21_1   chr7  HAVANA     CDS   \n",
       "141813   ENST00000684770.1_1  ENSG00000170379.21_1   chr7  HAVANA     CDS   \n",
       "141814   ENST00000684770.1_1  ENSG00000170379.21_1   chr7  HAVANA     CDS   \n",
       "141815   ENST00000684770.1_1  ENSG00000170379.21_1   chr7  HAVANA     CDS   \n",
       "141816   ENST00000684770.1_1  ENSG00000170379.21_1   chr7  HAVANA     CDS   \n",
       "\n",
       "            start        end score strand frame  \\\n",
       "0       127229137  127229217     .      +     2   \n",
       "1       127229539  127229648     .      +     2   \n",
       "2       127230120  127230191     .      +     0   \n",
       "3       127231017  127231142     .      +     0   \n",
       "4        64081423   64081539     .      +     2   \n",
       "...           ...        ...   ...    ...   ...   \n",
       "141812  143416776  143417767     .      +     1   \n",
       "141813  143418374  143418471     .      +     2   \n",
       "141814  143419751  143419964     .      +     0   \n",
       "141815  143420304  143420543     .      +     2   \n",
       "141816  143421453  143421790     .      +     2   \n",
       "\n",
       "                                                attribute       gene_type  \\\n",
       "0       gene_id \"ENSG00000004059.11_1\"; transcript_id ...  protein_coding   \n",
       "1       gene_id \"ENSG00000004059.11_1\"; transcript_id ...  protein_coding   \n",
       "2       gene_id \"ENSG00000004059.11_1\"; transcript_id ...  protein_coding   \n",
       "3       gene_id \"ENSG00000004059.11_1\"; transcript_id ...  protein_coding   \n",
       "4       gene_id \"ENSG00000173153.17_1\"; transcript_id ...  protein_coding   \n",
       "...                                                   ...             ...   \n",
       "141812  gene_id \"ENSG00000170379.21_1\"; transcript_id ...  protein_coding   \n",
       "141813  gene_id \"ENSG00000170379.21_1\"; transcript_id ...  protein_coding   \n",
       "141814  gene_id \"ENSG00000170379.21_1\"; transcript_id ...  protein_coding   \n",
       "141815  gene_id \"ENSG00000170379.21_1\"; transcript_id ...  protein_coding   \n",
       "141816  gene_id \"ENSG00000170379.21_1\"; transcript_id ...  protein_coding   \n",
       "\n",
       "                  exon_id  \n",
       "0       ENSE00003494180.1  \n",
       "1       ENSE00003504066.1  \n",
       "2       ENSE00003678978.1  \n",
       "3       ENSE00003676786.1  \n",
       "4       ENSE00003589560.1  \n",
       "...                   ...  \n",
       "141812  ENSE00003734225.1  \n",
       "141813  ENSE00002528833.1  \n",
       "141814  ENSE00002533605.1  \n",
       "141815  ENSE00002443125.1  \n",
       "141816  ENSE00002460631.1  \n",
       "\n",
       "[141817 rows x 13 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gencode_nover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "620b2067",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 32 s, sys: 4.64 s, total: 36.6 s\n",
      "Wall time: 36.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "#this is creating every possible position within 100 bp of each remaining CDS\n",
    "#making sure positions are unique in the case of two CDS within 100 bp of each other\n",
    "all_pos = list( set( [ ';'.join( [ chrom, str( pos ), strand, gene_id ] )\n",
    "                        for chrom, ( strand, gene_id ), ( start, end ) in zip( gencode_nover.chrom, zip( gencode_nover.strand, gencode_nover.transcript_id ), zip( gencode_nover.start, gencode_nover.end ) )\n",
    "                        for pos in range( start - 100, end + 101 ) ] ) )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b3be3582",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed( 212 )\n",
    "\n",
    "random_500k = random.choices( all_pos, k = 500000 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "30e1d5cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pos_count = Counter( random_500k )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "1d189a82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('chr8;30699626;-;ENST00000643185.2_1', 3),\n",
       " ('chr19;45207533;+;ENST00000587331.7_1', 3),\n",
       " ('chr6;30610508;+;ENST00000330083.6_1', 3),\n",
       " ('chr15;48818305;-;ENST00000316623.10_1', 3),\n",
       " ('chr18;55364826;-;ENST00000648908.2_1', 3),\n",
       " ('chr19;19427185;-;ENST00000247001.10_1', 3),\n",
       " ('chr18;54385372;+;ENST00000254442.8_1', 3),\n",
       " ('chr5;173372204;+;ENST00000265085.10_1', 3),\n",
       " ('chr13;96676044;-;ENST00000376747.8_1', 3),\n",
       " ('chr1;38463675;-;ENST00000373016.4_1', 3)]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#nice! All under 3!\n",
    "pos_count.most_common( 10 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "cae1d550",
   "metadata": {},
   "outputs": [],
   "source": [
    "assert pos_count.most_common( 1 )[ 0 ][ 1 ] <= 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "4e444609",
   "metadata": {},
   "outputs": [],
   "source": [
    "used_alt = { var: None for var in random_500k }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "512c7723",
   "metadata": {},
   "outputs": [],
   "source": [
    "by_chrom_fa = '/nfs/kitzman2/lab_common2/refs/human/ucsc_hg19/bychrom/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "31bf6509",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_seqs = { chrom: pp.get_refseq( by_chrom_fa + chrom + '.fa' )[ chrom ]\n",
    "             for chrom in gencode_nover.chrom.unique() }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "30c1b277",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_random_tbv( random_pos_list,\n",
    "                       refseqs,\n",
    "                       random_seed = 212 ):\n",
    "    \n",
    "    outd = { 'chrom': [],\n",
    "             'hg19_pos': [],\n",
    "             'ref': [],\n",
    "             'alt': [],\n",
    "             'strand': [],\n",
    "             'transcript_id': [] }\n",
    "    \n",
    "    used_alt = { var: [] for var in random_pos_list }\n",
    "    \n",
    "    alts = { 'A', 'C', 'G', 'T' }\n",
    "    \n",
    "    random.seed( random_seed )\n",
    "    \n",
    "    for var in random_pos_list:\n",
    "        \n",
    "        chrom, pos, strand, trans = var.split( ';' )\n",
    "        \n",
    "        outd[ 'chrom' ].append( chrom )\n",
    "        outd[ 'hg19_pos' ].append( int( pos ) )\n",
    "        outd[ 'ref' ].append( refseqs[ chrom ][ int( pos ) - 1 ].upper() )\n",
    "        #this is a mess but I'm making sure the alt is not the ref and that its not been used before\n",
    "        #I already know none of my variant positions occur >3 times - want to check before inputting\n",
    "        #I create a list with the ref and the alts already used - do a set difference from all possible alts and select\n",
    "        outd[ 'alt' ].append( random.choice( list( alts - set( used_alt[ var ] + [ outd[ 'ref' ][ -1 ] ] ) ) ) )\n",
    "        used_alt[ var ].append( outd[ 'alt' ][ -1 ] )\n",
    "        outd[ 'strand' ].append( strand )\n",
    "        outd[ 'transcript_id' ].append( trans )\n",
    "        \n",
    "    outdf = pd.DataFrame( outd ).sort_values( [ 'chrom', 'hg19_pos', 'strand' ] )\n",
    "    \n",
    "    return outdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "0d8aa356",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 12.2 s, sys: 185 ms, total: 12.4 s\n",
      "Wall time: 12.3 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "random_500k_df = create_random_tbv( random_500k,\n",
    "                                    ref_seqs )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "898a1af6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>hg19_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>251992</th>\n",
       "      <td>chr1</td>\n",
       "      <td>865446</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000616016.5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>360338</th>\n",
       "      <td>chr1</td>\n",
       "      <td>865560</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000616016.5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379094</th>\n",
       "      <td>chr1</td>\n",
       "      <td>865562</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000616016.5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6128</th>\n",
       "      <td>chr1</td>\n",
       "      <td>865640</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000616016.5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477545</th>\n",
       "      <td>chr1</td>\n",
       "      <td>865657</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000616016.5_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196062</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59252551</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000286448.12_1_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164131</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59252570</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000286448.12_1_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259776</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59252633</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000286448.12_1_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>164900</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59272476</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000286448.12_1_PAR_Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70835</th>\n",
       "      <td>chrY</td>\n",
       "      <td>59272513</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000286448.12_1_PAR_Y</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>500000 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom  hg19_pos ref alt strand               transcript_id\n",
       "251992  chr1    865446   G   A      +         ENST00000616016.5_1\n",
       "360338  chr1    865560   G   T      +         ENST00000616016.5_1\n",
       "379094  chr1    865562   A   C      +         ENST00000616016.5_1\n",
       "6128    chr1    865640   G   T      +         ENST00000616016.5_1\n",
       "477545  chr1    865657   T   A      +         ENST00000616016.5_1\n",
       "...      ...       ...  ..  ..    ...                         ...\n",
       "196062  chrY  59252551   G   C      +  ENST00000286448.12_1_PAR_Y\n",
       "164131  chrY  59252570   A   C      +  ENST00000286448.12_1_PAR_Y\n",
       "259776  chrY  59252633   T   G      +  ENST00000286448.12_1_PAR_Y\n",
       "164900  chrY  59272476   C   A      +  ENST00000286448.12_1_PAR_Y\n",
       "70835   chrY  59272513   A   T      +  ENST00000286448.12_1_PAR_Y\n",
       "\n",
       "[500000 rows x 6 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random_500k_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "1750a870",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_variant_cat( tbl_by_var,\n",
    "                         trans_gtf ):\n",
    "    \n",
    "    tbv = tbl_by_var.sort_values( by = 'transcript_id' ).copy()\n",
    "    gtf = trans_gtf.copy()\n",
    "    \n",
    "    exon = []\n",
    "    nearest_bd = []\n",
    "    variant_cat = []\n",
    "    \n",
    "    for trans, trans_tbv in tbv.groupby( 'transcript_id' ):\n",
    "        \n",
    "        trans_gtf = gtf.loc[ gtf.transcript_id == trans ].copy() \n",
    "        \n",
    "        for pos in trans_tbv.hg19_pos:\n",
    "            \n",
    "            exon.append( any( s <= pos and pos <= e for s,e in zip( trans_gtf.start, trans_gtf.end ) ) )\n",
    "            nearest_bd.append( np.min( [ abs( pos - bd ) \n",
    "                                         for bd in trans_gtf.start.tolist() + trans_gtf.end.tolist() ] ) )\n",
    "    \n",
    "            if exon[ -1 ]:\n",
    "            \n",
    "                if nearest_bd[ -1 ] < 10:\n",
    "                    variant_cat.append( 'Exonic Near SS' )\n",
    "                \n",
    "                else:\n",
    "                    variant_cat.append( 'Deep Exonic' )\n",
    "                \n",
    "            else:\n",
    "        \n",
    "                if nearest_bd[ -1 ] <= 2:\n",
    "                    variant_cat.append( 'Canonical' )\n",
    "            \n",
    "                elif nearest_bd[ -1 ] <= 10:\n",
    "                    variant_cat.append( 'Intronic Near SS' )\n",
    "                \n",
    "                elif nearest_bd[ -1 ] <= 100:\n",
    "                    variant_cat.append( 'Deep Intronic' )\n",
    "                    \n",
    "                else:\n",
    "                    variant_cat.append( np.nan )\n",
    "                    \n",
    "    tbv[ 'exon' ] = exon\n",
    "    tbv[ 'variant_cat' ] = variant_cat\n",
    "    \n",
    "    return tbv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "ae47a6f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2min 54s, sys: 283 ms, total: 2min 54s\n",
      "Wall time: 2min 54s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "random_500k_df = compute_variant_cat( random_500k_df, \n",
    "                                      gencode_nover )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ae72f3c0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Deep Intronic': 50.544599999999996,\n",
       " 'Deep Exonic': 37.6914,\n",
       " 'Intronic Near SS': 4.7812,\n",
       " 'Canonical': 1.166,\n",
       " 'Exonic Near SS': 5.8168}"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{ cat: 100*( ( random_500k_df.variant_cat == cat ).sum() / len( random_500k_df ) )\n",
    "  for cat in random_500k_df.variant_cat.unique() }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "d2ba65bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_gtf = gencode_annot.set_index( 'gene_id' ).loc[ gencode_nover.gene_id.unique() ].reset_index()[ [ 'chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute' ] ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "f4ebf718",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_gtf.to_csv( annots_dir + 'gencode/gencode.v38lift37.random_vcf.one_trans_500k.gtf.gz',\n",
    "                sep = '\\t',\n",
    "                index = False,\n",
    "                header = False,\n",
    "                quoting = csv.QUOTE_NONE,\n",
    "                compression = 'gzip' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "8f1c13b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "splai_onetrans = css.gtf_to_splai_annot( out_gtf )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "312e503c",
   "metadata": {},
   "outputs": [],
   "source": [
    "splai_onetrans.to_csv( annots_dir + 'gencode/gencode.v38lift37.random_vcf.one_trans_500k.txt',\n",
    "                       sep = '\\t',\n",
    "                       index = False )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "2eb2ad11",
   "metadata": {},
   "outputs": [],
   "source": [
    "random_500k_df.to_csv( '/nfs/kitzman2/smithcat/proj/spliceAI/input_vcfs/random_vars_500k.txt',\n",
    "                         sep = '\\t',\n",
    "                         index = False )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "c792ae1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_out = '/nfs/kitzman2/smithcat/proj/spliceAI/input_vcfs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "5f8567ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf.create_input_vcf( vcf.vcf_header(),\n",
    "                      vcf.tup_to_vcf( [ ( chrom[ 3: ], pos, ref, alt )\n",
    "                                        for ( chrom, pos ), ( ref, alt ) \n",
    "                                        in zip( zip( random_500k_df.chrom, random_500k_df.hg19_pos ), \n",
    "                                                zip( random_500k_df.ref, random_500k_df.alt ) ) ] ),\n",
    "                      vcf_out + 'random_vars_500k.vcf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "40d25cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf.create_input_vcf( vcf.vcf_header_chr(),\n",
    "                      vcf.tup_to_vcf( [ ( chrom, pos, ref, alt )\n",
    "                                        for ( chrom, pos ), ( ref, alt ) \n",
    "                                        in zip( zip( random_500k_df.chrom, random_500k_df.hg19_pos ), \n",
    "                                                zip( random_500k_df.ref, random_500k_df.alt ) ) ] ),\n",
    "                      vcf_out + 'random_vars_500k_chr.vcf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "50c38764",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf.create_input_vcf( vcf.vcf_header(),\n",
    "                      vcf.tup_to_vcf( [ ( chrom[ 3: ], pos, ref_seqs[ chrom ][ int( pos ) - 1 ], alt )\n",
    "                                        if ref_seqs[ chrom ][ int( pos ) - 1 ].isupper()\n",
    "                                        else ( chrom[ 3: ], pos, ref_seqs[ chrom ][ int( pos ) - 1 ], alt.lower() )\n",
    "                                        for ( chrom, pos ), ( ref, alt ) \n",
    "                                        in zip( zip( random_500k_df.chrom, random_500k_df.hg19_pos ), \n",
    "                                                zip( random_500k_df.ref, random_500k_df.alt ) ) ] ),\n",
    "                      vcf_out + 'random_vars_500k_lowercase.vcf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "2c168f4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>hg19_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>exon</th>\n",
       "      <th>variant_cat</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>429518</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127229451</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181682</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127229679</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58524</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127229473</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397918</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127229245</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138290</th>\n",
       "      <td>chr7</td>\n",
       "      <td>127231224</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000000233.10_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234550</th>\n",
       "      <td>chr7</td>\n",
       "      <td>117305687</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000003084.11_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>249693</th>\n",
       "      <td>chr7</td>\n",
       "      <td>117243886</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000003084.11_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447047</th>\n",
       "      <td>chr7</td>\n",
       "      <td>117293067</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000003084.11_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87062</th>\n",
       "      <td>chr7</td>\n",
       "      <td>117293017</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000003084.11_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>296503</th>\n",
       "      <td>chr7</td>\n",
       "      <td>117293010</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000003084.11_1</td>\n",
       "      <td>False</td>\n",
       "      <td>Deep Intronic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom   hg19_pos ref alt strand         transcript_id   exon  \\\n",
       "429518  chr7  127229451   G   A      +  ENST00000000233.10_1  False   \n",
       "181682  chr7  127229679   C   A      +  ENST00000000233.10_1  False   \n",
       "58524   chr7  127229473   T   G      +  ENST00000000233.10_1  False   \n",
       "397918  chr7  127229245   G   C      +  ENST00000000233.10_1  False   \n",
       "138290  chr7  127231224   T   G      +  ENST00000000233.10_1  False   \n",
       "...      ...        ...  ..  ..    ...                   ...    ...   \n",
       "234550  chr7  117305687   T   C      +  ENST00000003084.11_1  False   \n",
       "249693  chr7  117243886   G   C      +  ENST00000003084.11_1  False   \n",
       "447047  chr7  117293067   C   A      +  ENST00000003084.11_1  False   \n",
       "87062   chr7  117293017   G   T      +  ENST00000003084.11_1  False   \n",
       "296503  chr7  117293010   T   G      +  ENST00000003084.11_1  False   \n",
       "\n",
       "          variant_cat  \n",
       "429518  Deep Intronic  \n",
       "181682  Deep Intronic  \n",
       "58524   Deep Intronic  \n",
       "397918  Deep Intronic  \n",
       "138290  Deep Intronic  \n",
       "...               ...  \n",
       "234550  Deep Intronic  \n",
       "249693  Deep Intronic  \n",
       "447047  Deep Intronic  \n",
       "87062   Deep Intronic  \n",
       "296503  Deep Intronic  \n",
       "\n",
       "[100 rows x 8 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random_500k_df.loc[ random_500k_df.variant_cat == 'Deep Intronic' ][ : 100 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "d2a3e402",
   "metadata": {},
   "outputs": [],
   "source": [
    "smol_deep_intronic = random_500k_df.loc[ random_500k_df.variant_cat == 'Deep Intronic' ][ : 100 ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "2524d852",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf.create_input_vcf( vcf.vcf_header(),\n",
    "                      vcf.tup_to_vcf( [ ( chrom[ 3: ], pos, ref, alt )\n",
    "                                        for ( chrom, pos ), ( ref, alt ) \n",
    "                                        in zip( zip( smol_deep_intronic.chrom, smol_deep_intronic.hg19_pos ), \n",
    "                                                zip( smol_deep_intronic.ref, smol_deep_intronic.alt ) ) ] ),\n",
    "                      vcf_out + 'random_vars_500k_smol_dint.vcf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "00252f08",
   "metadata": {},
   "outputs": [],
   "source": [
    "smol_randoms = random_500k_df[ : 100 ].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "8b8495f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf.create_input_vcf( vcf.vcf_header(),\n",
    "                      vcf.tup_to_vcf( [ ( chrom[ 3: ], pos, ref, alt )\n",
    "                                        for ( chrom, pos ), ( ref, alt ) \n",
    "                                        in zip( zip( smol_randoms.chrom, smol_randoms.hg19_pos ), \n",
    "                                                zip( smol_randoms.ref, smol_randoms.alt ) ) ] ),\n",
    "                      vcf_out + 'random_vars_500k_smol_random_set.vcf' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "395caed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = pp.get_refseq( '/nfs/kitzman2/lab_common2/refs/human/ucsc_hg19/hg19_allcaps.fa' )[ 'chr11' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "28917097",
   "metadata": {},
   "outputs": [],
   "source": [
    "b = pp.get_refseq( '/nfs/kitzman2/lab_common2/refs/human/ucsc_hg19/hg19.fa' )[ 'chr11' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6fd0011-e004-4165-b2ee-7591ed25f4bb",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "a8d8e6f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "ed837a36",
   "metadata": {},
   "outputs": [],
   "source": [
    "import splanl.score_motifs as sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "ae3f3fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "smol_vcf = pysam.VariantFile( '/nfs/kitzman2/smithcat/proj/spliceAI/pangolin_output/random_vars_500k_smol_random_set.vcf',\n",
    "                                      mode = 'r' )\n",
    "\n",
    "pang_smol = sm.get_pangolin_scores( smol_vcf )\n",
    "\n",
    "smol_vcf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2b76e19a",
   "metadata": {},
   "outputs": [],
   "source": [
    "lorge_vcf = pysam.VariantFile( '/nfs/kitzman2/smithcat/proj/spliceAI/pangolin_output/random_vars_500k.vcf',\n",
    "                                      mode = 'r' )\n",
    "\n",
    "pang_lorge = sm.get_pangolin_scores( lorge_vcf )\n",
    "\n",
    "lorge_vcf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "00aac26e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>hg19_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>pang_incr</th>\n",
       "      <th>pang_incr_pos</th>\n",
       "      <th>pang_decr</th>\n",
       "      <th>pang_decr_pos</th>\n",
       "      <th>pang_max</th>\n",
       "      <th>pang_max_type</th>\n",
       "      <th>pang_max_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>127229451</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-117</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-117</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>127231031</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-8</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>272</td>\n",
       "      <td>0.03</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>127230172</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00</td>\n",
       "      <td>19</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>127229679</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>127229569</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.01</td>\n",
       "      <td>15</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.01</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>2</td>\n",
       "      <td>37463349</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-30</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>2</td>\n",
       "      <td>37473279</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>2</td>\n",
       "      <td>37459259</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-10</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>150</td>\n",
       "      <td>0.01</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>2</td>\n",
       "      <td>37473395</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-63</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-63</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>2</td>\n",
       "      <td>37474590</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>52</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>pang_decr</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   hg19_pos ref alt  pang_incr  pang_incr_pos  pang_decr  \\\n",
       "0      7  127229451   G   A       0.00           -117      -0.00   \n",
       "1      7  127231031   G   A       0.03             -8      -0.00   \n",
       "2      7  127230172   C   T       0.00             19      -0.00   \n",
       "3      7  127229679   C   A       0.00             10      -0.00   \n",
       "4      7  127229569   A   T       0.01             15      -0.00   \n",
       "..   ...        ...  ..  ..        ...            ...        ...   \n",
       "95     2   37463349   C   T       0.00            -30      -0.00   \n",
       "96     2   37473279   G   C       0.00              3      -0.00   \n",
       "97     2   37459259   T   A       0.01            -10      -0.00   \n",
       "98     2   37473395   A   C       0.00            -63      -0.00   \n",
       "99     2   37474590   A   C       0.01              9      -0.21   \n",
       "\n",
       "    pang_decr_pos  pang_max pang_max_type  pang_max_pos  \n",
       "0            -117      0.00     pang_incr          -117  \n",
       "1             272      0.03     pang_incr            -8  \n",
       "2             -29      0.00     pang_incr            19  \n",
       "3             -95      0.00     pang_incr            10  \n",
       "4             -30      0.01     pang_incr            15  \n",
       "..            ...       ...           ...           ...  \n",
       "95            -30      0.00     pang_incr           -30  \n",
       "96             59      0.00     pang_incr             3  \n",
       "97            150      0.01     pang_incr           -10  \n",
       "98             29      0.00     pang_incr           -63  \n",
       "99             52     -0.21     pang_decr            52  \n",
       "\n",
       "[100 rows x 11 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pang_smol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "49dd53f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "pang_smol = pang_smol.rename( columns = { col: col + '_smol' for col in pang_smol if col.startswith( 'pang_' ) } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "80586bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx_cols = [ 'chrom', 'hg19_pos', 'ref', 'alt' ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "05bc172d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pang = pang_smol.set_index( idx_cols ).merge( pang_lorge.set_index( idx_cols ),\n",
    "                                              how = 'left', \n",
    "                                              left_index = True,\n",
    "                                              right_index = True ).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "c6333cd9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>hg19_pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>pang_incr_smol</th>\n",
       "      <th>pang_incr_pos_smol</th>\n",
       "      <th>pang_decr_smol</th>\n",
       "      <th>pang_decr_pos_smol</th>\n",
       "      <th>pang_max_smol</th>\n",
       "      <th>pang_max_type_smol</th>\n",
       "      <th>pang_max_pos_smol</th>\n",
       "      <th>pang_incr</th>\n",
       "      <th>pang_incr_pos</th>\n",
       "      <th>pang_decr</th>\n",
       "      <th>pang_decr_pos</th>\n",
       "      <th>pang_max</th>\n",
       "      <th>pang_max_type</th>\n",
       "      <th>pang_max_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>127229451</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-117</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-117</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-117</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-3</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>88</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7</td>\n",
       "      <td>127231031</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.03</td>\n",
       "      <td>-8</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>272</td>\n",
       "      <td>0.03</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-8</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-8</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-14</td>\n",
       "      <td>0.01</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>7</td>\n",
       "      <td>127230172</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00</td>\n",
       "      <td>19</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>19</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-2</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>19</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>127229679</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>10</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-31</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>127229569</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.01</td>\n",
       "      <td>15</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.01</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>15</td>\n",
       "      <td>0.00</td>\n",
       "      <td>15</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>2</td>\n",
       "      <td>37463349</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-30</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-30</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-299</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>2</td>\n",
       "      <td>37473279</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>59</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>3</td>\n",
       "      <td>0.01</td>\n",
       "      <td>3</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-84</td>\n",
       "      <td>0.01</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>2</td>\n",
       "      <td>37459259</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>0.01</td>\n",
       "      <td>-10</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>150</td>\n",
       "      <td>0.01</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-10</td>\n",
       "      <td>0.00</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-300</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>2</td>\n",
       "      <td>37473395</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-63</td>\n",
       "      <td>-0.00</td>\n",
       "      <td>29</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-63</td>\n",
       "      <td>0.00</td>\n",
       "      <td>2</td>\n",
       "      <td>-0.0</td>\n",
       "      <td>-200</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>2</td>\n",
       "      <td>37474590</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.01</td>\n",
       "      <td>9</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>52</td>\n",
       "      <td>-0.21</td>\n",
       "      <td>pang_decr</td>\n",
       "      <td>52</td>\n",
       "      <td>0.00</td>\n",
       "      <td>-79</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-300</td>\n",
       "      <td>0.00</td>\n",
       "      <td>pang_incr</td>\n",
       "      <td>-79</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   hg19_pos ref alt  pang_incr_smol  pang_incr_pos_smol  \\\n",
       "0      7  127229451   G   A            0.00                -117   \n",
       "1      7  127231031   G   A            0.03                  -8   \n",
       "2      7  127230172   C   T            0.00                  19   \n",
       "3      7  127229679   C   A            0.00                  10   \n",
       "4      7  127229569   A   T            0.01                  15   \n",
       "..   ...        ...  ..  ..             ...                 ...   \n",
       "95     2   37463349   C   T            0.00                 -30   \n",
       "96     2   37473279   G   C            0.00                   3   \n",
       "97     2   37459259   T   A            0.01                 -10   \n",
       "98     2   37473395   A   C            0.00                 -63   \n",
       "99     2   37474590   A   C            0.01                   9   \n",
       "\n",
       "    pang_decr_smol  pang_decr_pos_smol  pang_max_smol pang_max_type_smol  \\\n",
       "0            -0.00                -117           0.00          pang_incr   \n",
       "1            -0.00                 272           0.03          pang_incr   \n",
       "2            -0.00                 -29           0.00          pang_incr   \n",
       "3            -0.00                 -95           0.00          pang_incr   \n",
       "4            -0.00                 -30           0.01          pang_incr   \n",
       "..             ...                 ...            ...                ...   \n",
       "95           -0.00                 -30           0.00          pang_incr   \n",
       "96           -0.00                  59           0.00          pang_incr   \n",
       "97           -0.00                 150           0.01          pang_incr   \n",
       "98           -0.00                  29           0.00          pang_incr   \n",
       "99           -0.21                  52          -0.21          pang_decr   \n",
       "\n",
       "    pang_max_pos_smol  pang_incr  pang_incr_pos  pang_decr  pang_decr_pos  \\\n",
       "0                -117       0.00             -3       -0.0             88   \n",
       "1                  -8       0.01             -8       -0.0            -14   \n",
       "2                  19       0.00             -2       -0.0             19   \n",
       "3                  10       0.00             10       -0.0            -31   \n",
       "4                  15       0.00             15       -0.0            -30   \n",
       "..                ...        ...            ...        ...            ...   \n",
       "95                -30       0.00             -2        0.0           -299   \n",
       "96                  3       0.01              3       -0.0            -84   \n",
       "97                -10       0.00              7        0.0           -300   \n",
       "98                -63       0.00              2       -0.0           -200   \n",
       "99                 52       0.00            -79        0.0           -300   \n",
       "\n",
       "    pang_max pang_max_type  pang_max_pos  \n",
       "0       0.00     pang_incr            -3  \n",
       "1       0.01     pang_incr            -8  \n",
       "2       0.00     pang_incr            -2  \n",
       "3       0.00     pang_incr            10  \n",
       "4       0.00     pang_incr            15  \n",
       "..       ...           ...           ...  \n",
       "95      0.00     pang_incr            -2  \n",
       "96      0.01     pang_incr             3  \n",
       "97      0.00     pang_incr             7  \n",
       "98      0.00     pang_incr             2  \n",
       "99      0.00     pang_incr           -79  \n",
       "\n",
       "[100 rows x 18 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pang"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "88039788-8d2d-49cd-bb70-95c70d89def7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "a7e8fb05",
   "metadata": {},
   "outputs": [],
   "source": [
    "smol_vcf = pysam.VariantFile( '/nfs/kitzman2/smithcat/proj/spliceAI/pangolin_output/random_vars_500k_smol_random_set_gl.vcf',\n",
    "                                      mode = 'r' )\n",
    "\n",
    "pang_smol_gl = sm.get_pangolin_scores( smol_vcf )\n",
    "\n",
    "smol_vcf.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "6a35f5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "pang_smol_gl = pang_smol_gl.rename( columns = { col: col + '_smol_gl' for col in pang_smol_gl if col.startswith( 'pang_' ) } )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "a8ac9058",
   "metadata": {},
   "outputs": [],
   "source": [
    "pang = pang.set_index( idx_cols ).merge( pang_smol_gl.set_index( idx_cols ),\n",
    "                                              how = 'left', \n",
    "                                              left_index = True,\n",
    "                                              right_index = True ).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83df0dd4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
